pkg/sql/schemachanger/scbuild/internal/scbuildstmt/create_index.go

```diff
@@ -350,13 +351,23 @@ func addColumnsForSecondaryIndex(
 
 	// Check that the index creation spec is sane.
 	columnRefs := map[string]struct{}{}
+	columnExprRefs := map[string]struct{}{}
 	for _, columnNode := range n.Columns {
-		colName := columnNode.Column.String()
-		if _, found := columnRefs[colName]; found {
-			panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
-				"index %q contains duplicate column %q", n.Name, colName))
+		if columnNode.Expr == nil {
+			colName := columnNode.Column.Normalize()
+			if _, found := columnRefs[colName]; found {
+				panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
+					"index %q contains duplicate column %q", n.Name, colName))
+			}
+			columnRefs[colName] = struct{}{}
+		} else {
+			colExpr := columnNode.Expr.String()
+			if _, found := columnExprRefs[colExpr]; found {
+				panic(pgerror.Newf(pgcode.InvalidObjectDefinition,
+					"index %q contains duplicate expression", n.Name))
+			}
+			columnExprRefs[colExpr] = struct{}{}
```

Comment:
Yes, those are the two cases, pretty much either the same column referenced twice or the same expression twice since neither makes sense. I'll add a comment:

// Detect if either the same column or expression repeats twice in this
// index definition. ie.:
// 1) CREATE INDEX idx ON t(i, i) or
// 2) CREATE INDEX idx ON t(lower(i), j, lower(i)).